tiatoolbox/models/architecture/sccnn.py

```diff
@@ -0,0 +1,372 @@
+                kernel_size (int):
+                    Kernel size of convolution filter.
+
+            Returns:
+                torch.nn.ModuleDict:
+                    Module dictionary.
+
+            """
+            module_dict = OrderedDict()
+            module_dict["conv1"] = nn.Sequential(
+                nn.Conv2d(
+                    in_channels,
+                    out_channels,
+                    kernel_size=(kernel_size, kernel_size),
+                    stride=(1, 1),
+                    padding=0,
+                    bias=True,
+                ),
+                nn.ReLU(),
+            )
+
+            return nn.ModuleDict(module_dict)
+
+        def spatially_constrained_layer1(
+            in_channels, out_channels
+        ) -> torch.nn.ModuleDict:
+            """Spatially constrained layer.
+
+            Takes fully connected layer and returns outputs for creating probability
+            map for the output.
+
+            Args:
+                in_channels (int):
+                    Number of channels in input.
+                out_channels (int):
+                    Number of required channels in output.
+
+            Returns:
+                torch.nn.ModuleDict:
+                    Module dictionary.
+
+            """
+            module_dict = OrderedDict()
+            module_dict["conv1"] = nn.Sequential(
+                nn.Conv2d(
+                    in_channels,
+                    out_channels,
+                    kernel_size=(1, 1),
+                    stride=(1, 1),
+                    padding=0,
+                    bias=True,
+                ),
+                nn.Sigmoid(),
+            )
+            return nn.ModuleDict(module_dict)
+
+        module_dict = OrderedDict()
+        module_dict["l1"] = conv_act_branch(num_input_channels, 30, 2)
+        module_dict["pool1"] = nn.MaxPool2d(2, padding=0)
+        module_dict["l2"] = conv_act_branch(30, 60, 2)
+        module_dict["pool2"] = nn.MaxPool2d(2, padding=0)
+        module_dict["l3"] = conv_act_branch(60, 90, 3)
+        module_dict["l4"] = conv_act_branch(90, 1024, 5)
+        module_dict["dropout1"] = nn.Dropout2d(p=0.5)
+        module_dict["l5"] = conv_act_branch(1024, 512, 1)
+        module_dict["dropout2"] = nn.Dropout2d(p=0.5)
+        module_dict["sc"] = spatially_constrained_layer1(512, 3)
+
+        self.layer = nn.ModuleDict(module_dict)
+
+    def forward(self, input_tensor: torch.Tensor):  # skipcq: PYL-W0221
+        """Logic for using layers defined in init.
+
+        This method defines how layers are used in forward operation.
+
+        Args:
+            input_tensor (torch.Tensor):
+                Input images, the tensor is in the shape of NCHW.
+
+        Returns:
+            torch.Tensor:
+                Output map for cell detection. Peak detection should be applied
+                to this output for cell detection.
+
+        """
+
+        def conv_act_branch(
```

Comment:
I don't see a reason to define this single operation function with the same `conv_act_branch`  name here. Instead, L272 can be rewritten like this: `l1 = self.layer["l1"](input_tensor)`

Reply:
Same as mostafa, if it is just a singular function call then no need to do this, just directly calling the layer.

```python
conv_act_branch(self.layer["l4"], in_tensor=l3)
```
to
```
self.layer["l4"](l3)
```

this wont effect the variable weights at all as nothing is rewritten in `__init__`, no need to retrain

Reply:
I get this error, if replace the code with above suggestion.


self = ModuleDict(
  (conv1): Sequential(
    (0): Conv2d(3, 30, kernel_size=(2, 2), stride=(1, 1))
    (1): ReLU()
  )
)
input = (tensor([[[[0.8431, 0.8549, 0.8549,  ..., 0.8510, 0.8941, 0.9412],
          [0.8118, 0.8588, 0.8863,  ..., 0.8588, 0...., 0.7333, 0.7412,  ..., 0.5765, 0.6078, 0.6471],
          [0.7294, 0.7412, 0.7529,  ..., 0.6314, 0.6471, 0.6706]]]]),)

    def _forward_unimplemented(self, *input: Any) -> None:
        r"""Defines the computation performed at every call.
    
        Should be overridden by all subclasses.
    
        .. note::
            Although the recipe for forward pass needs to be defined within
            this function, one should call the :class:`Module` instance afterwards
            instead of this since the former takes care of running the
            registered hooks while the latter silently ignores them.
        """
>       raise NotImplementedError
E       NotImplementedError

C:\tools\Anaconda3\envs\tiatoolbox-dev\lib\site-packages\torch\nn\modules\module.py:201: NotImplementedError

Reply:
Removed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/434/commits/3ab3362d1463507df3a953cf5ee1da4a02af2f41